pkg/extensions/sync/destination.go

```diff
@@ -80,178 +81,264 @@ func (registry *DestinationRegistry) CanSkipImage(repo, tag string, imageDigest
 	return registry.tempStorage.GetImageReference(repo, reference)
 }
 
 // finalize a syncing image.
-func (registry *DestinationRegistry) CommitImage(imageReference types.ImageReference, repo, reference string) error {
+func (registry *DestinationRegistry) CommitAll(repo string, imageReference ref.Ref) error {
 	imageStore := registry.storeController.GetImageStore(repo)
 
-	tempImageStore := getImageStoreFromImageReference(imageReference, repo, reference, registry.log)
+	tempImageStore := getImageStoreFromImageReference(repo, imageReference, registry.log)
 
 	defer os.RemoveAll(tempImageStore.RootDir())
 
-	registry.log.Info().Str("syncTempDir", path.Join(tempImageStore.RootDir(), repo)).Str("reference", reference).
+	registry.log.Info().Str("syncTempDir", path.Join(tempImageStore.RootDir(), repo)).Str("repository", repo).
 		Msg("pushing synced local image to local registry")
 
 	var lockLatency time.Time
 
-	manifestBlob, manifestDigest, mediaType, err := tempImageStore.GetImageManifest(repo, reference)
+	index, err := storageCommon.GetIndex(tempImageStore, repo, registry.log)
 	if err != nil {
 		registry.log.Error().Str("errorType", common.TypeOf(err)).
-			Err(err).Str("dir", path.Join(tempImageStore.RootDir(), repo)).Str("repo", repo).Str("reference", reference).
-			Msg("couldn't find synced manifest in temporary sync dir")
+			Err(err).Str("dir", path.Join(tempImageStore.RootDir(), repo)).Str("repo", repo).
+			Msg("failed to get repo index from temp sync dir")
 
 		return err
 	}
 
-	// is image manifest
-	switch mediaType {
-	case ispec.MediaTypeImageManifest:
-		if err := registry.copyManifest(repo, manifestBlob, reference, tempImageStore); err != nil {
-			if errors.Is(err, zerr.ErrImageLintAnnotations) {
-				registry.log.Error().Str("errorType", common.TypeOf(err)).
-					Err(err).Msg("couldn't upload manifest because of missing annotations")
+	seen := []godigest.Digest{}
 
-				return nil
-			}
-
-			return err
-		}
-	case ispec.MediaTypeImageIndex:
-		// is image index
-		var indexManifest ispec.Index
+	for _, desc := range index.Manifests {
+		reference := GetDescriptorReference(desc)
 
-		if err := json.Unmarshal(manifestBlob, &indexManifest); err != nil {
+		manifestBlob, manifestDigest, mediaType, err := tempImageStore.GetImageManifest(repo, reference)
+		if err != nil {
 			registry.log.Error().Str("errorType", common.TypeOf(err)).
-				Err(err).Str("dir", path.Join(tempImageStore.RootDir(), repo)).
-				Msg("invalid JSON")
+				Err(err).Str("dir", path.Join(tempImageStore.RootDir(), repo)).Str("repo", repo).Str("reference", reference).
+				Msg("failed to get manifest from temporary sync dir")
 
 			return err
 		}
 
-		for _, manifest := range indexManifest.Manifests {
-			tempImageStore.RLock(&lockLatency)
-			manifestBuf, err := tempImageStore.GetBlobContent(repo, manifest.Digest)
-			tempImageStore.RUnlock(&lockLatency)
+		// is image manifest
```

Comment:
can't we call copyManifest here? It looks like the logic to check media type and copy is similar to the one in this function